File: studio/core/store.js
// The live twin state model. One store, filled by socket.js, read by every
// screen. Not a generic framework store — shaped directly around the three
// message types the hub actually emits (snapshot / point / event), see
// docs/APP_PLAN.md §0.

const HISTORY_LIMIT = 500;
const FEED_LIMIT = 80;
const SERIES_LIMIT = 300; // per key — this session only; see docs/APP_PLAN.md §7

function createStore() {
  const state = {
    connection: "connecting", // connecting | live | down
    now: 0,
    points: new Map(),        // key -> {value, t, source, confidence}
    series: new Map(),        // key -> [{t, value}, …] — bounded, numeric points only, this session
    actions: new Map(),       // id -> {id, cause, evidence, command, expect, status, retries, reason}
    actionOrder: [],          // most-recent-first ids
    feed: [],                 // recent {topic, payload, t}, most-recent-first
    scenario: null,
  };

  const listeners = new Set();
  function notify() { for (const fn of listeners) fn(state); }

  return {
    state,
    subscribe(fn) { listeners.add(fn); return () => listeners.delete(fn); },

    setConnection(status) { state.connection = status; notify(); },

    applySnapshot(msg) {
      state.now = msg.t || 0;
      state.points = new Map(Object.entries(msg.points || {}));
      for (const [key, p] of state.points) this._pushSeries(key, p.t, p.value);
      notify();
    },

    applyPoint(msg) {
      state.now = msg.t ?? state.now;
      state.points.set(msg.key, { value: msg.value, t: msg.t, source: msg.source, confidence: msg.confidence });
      this._pushSeries(msg.key, msg.t, msg.value);
      notify();
    },

    _pushSeries(key, t, value) {
      if (typeof value !== "number" || !Number.isFinite(value)) return;
      let arr = state.series.get(key);
      if (!arr) { arr = []; state.series.set(key, arr); }
      if (arr.length && arr[arr.length - 1].t === t) { arr[arr.length - 1].value = value; return; }
      arr.push({ t, value });
      if (arr.length > SERIES_LIMIT) arr.shift();
    },

    series(key) {
      return state.series.get(key) || [];
    },

    applyEvent(msg) {
      state.now = msg.t ?? state.now;
      state.feed.unshift({ topic: msg.topic, payload: msg.payload, t: msg.t });
      if (state.feed.length > FEED_LIMIT) state.feed.length = FEED_LIMIT;

      const payload = msg.payload || {};
      const a = payload.action;
      if ((msg.topic === "domora/plan/action" || msg.topic === "domora/act/dispatched") && a) {
        this._upsertAction(a);
      } else if (msg.topic === "domora/verify/confirmed") {
        this._markAction(payload.action_id, "confirmed");
      } else if (msg.topic.startsWith("domora/alert") && payload.action_id != null) {
        this._markAction(payload.action_id, "failed", { reason: payload.reason });
      }
      notify();
    },

    _upsertAction(a) {
      let m = state.actions.get(a.id);
      if (!m) {
        m = { id: a.id, status: "pending" };
        state.actions.set(a.id, m);
        state.actionOrder.unshift(a.id);
        if (state.actionOrder.length > HISTORY_LIMIT) state.actionOrder.length = HISTORY_LIMIT;
      }
      m.cause = a.cause ?? m.cause;
      m.command = a.command_topic ?? m.command;
      m.evidence = a.evidence ?? m.evidence;
      if (a.expectation) m.expect = a.expectation.describe;
      if (a.retries != null) m.retries = a.retries;
    },

    _markAction(id, status, extra) {
      let m = state.actions.get(id);
      if (!m) { m = { id, status: "pending" }; state.actions.set(id, m); state.actionOrder.unshift(id); }
      m.status = status;
      if (extra) Object.assign(m, extra);
    },

    point(key, dflt = undefined) {
      const p = state.points.get(key);
      return p ? p.value : dflt;
    },

    reset() {
      state.now = 0; state.points = new Map(); state.series = new Map(); state.actions = new Map();
      state.actionOrder = []; state.feed = [];
      notify();
    },
  };
}

export const store = createStore();

File: mobile/lib/core/store.dart
// The live twin state model — direct Dart port of studio/core/store.js.
// One store, filled by hub_client.dart's WebSocket handler, read by every
// screen via ChangeNotifier. Shaped around the three message types the hub
// actually emits (snapshot / point / event) — see docs/APP_PLAN.md §0.
import 'package:flutter/foundation.dart';

class Point {
  final dynamic value;
  final int t;
  final String? source;
  final double? confidence;
  const Point({required this.value, required this.t, this.source, this.confidence});
}

class SeriesPoint {
  final int t;
  final double value;
  const SeriesPoint(this.t, this.value);
}

class FeedEntry {
  final String topic;
  final Map<String, dynamic> payload;
  final int t;
  const FeedEntry({required this.topic, required this.payload, required this.t});
}

class ActionCard {
  final int id;
  String? cause;
  String? command;
  Map<String, dynamic>? evidence;
  String? expect;
  int retries = 0;
  String status = 'pending'; // pending | confirmed | failed
  String? reason;
  ActionCard(this.id);
}

class DomoraStore extends ChangeNotifier {
  static const _seriesLimit = 300; // per key — this session only, see APP_PLAN §7
  static const _feedLimit = 80;

  // unconfigured: no hub address set yet (or none reachable to attempt) —
  // the honest default before HubClient.connect() has ever been asked to
  // dial anything. connecting/live/down all imply an attempt is underway.
  String connection = 'unconfigured'; // unconfigured | connecting | live | down
  int now = 0;
  final Map<String, Point> points = {};
  final Map<String, List<SeriesPoint>> series = {};
  final Map<int, ActionCard> actions = {};
  final List<int> actionOrder = []; // most-recent-first
  final List<FeedEntry> feed = []; // most-recent-first

  void setConnection(String status) {
    connection = status;
    notifyListeners();
  }

  void applySnapshot(Map<String, dynamic> msg) {
    now = _asInt(msg['t']) ?? 0;
    points.clear();
    final pts = (msg['points'] as Map?)?.cast<String, dynamic>() ?? const {};
    pts.forEach((key, v) {
      final map = (v as Map).cast<String, dynamic>();
      final p = Point(
        value: map['value'],
        t: _asInt(map['t']) ?? now,
        source: map['source'] as String?,
        confidence: (map['confidence'] as num?)?.toDouble(),
      );
      points[key] = p;
      _pushSeries(key, p.t, p.value);
    });
    notifyListeners();
  }

  void applyPoint(Map<String, dynamic> msg) {
    now = _asInt(msg['t']) ?? now;
    final key = msg['key'] as String;
    final p = Point(
      value: msg['value'],
      t: now,
      source: msg['source'] as String?,
      confidence: (msg['confidence'] as num?)?.toDouble(),
    );
    points[key] = p;
    _pushSeries(key, p.t, p.value);
    notifyListeners();
  }

  void applyEvent(Map<String, dynamic> msg) {
    now = _asInt(msg['t']) ?? now;
    final topic = msg['topic'] as String;
    final payload = (msg['payload'] as Map?)?.cast<String, dynamic>() ?? const {};
    feed.insert(0, FeedEntry(topic: topic, payload: payload, t: now));
    if (feed.length > _feedLimit) feed.removeLast();

    final a = payload['action'];
    if ((topic == 'domora/plan/action' || topic == 'domora/act/dispatched') && a != null) {
      _upsertAction((a as Map).cast<String, dynamic>());
    } else if (topic == 'domora/verify/confirmed') {
      final id = _asInt(payload['action_id']);
      if (id != null) _markAction(id, 'confirmed');
    } else if (topic.startsWith('domora/alert')) {
      final id = _asInt(payload['action_id']);
      if (id != null) _markAction(id, 'failed', reason: payload['reason'] as String?);
    }
    notifyListeners();
  }

  void _pushSeries(String key, int t, dynamic value) {
    if (value is! num) return;
    final arr = series.putIfAbsent(key, () => []);
    if (arr.isNotEmpty && arr.last.t == t) {
      arr[arr.length - 1] = SeriesPoint(t, value.toDouble());
      return;
    }
    arr.add(SeriesPoint(t, value.toDouble()));
    if (arr.length > _seriesLimit) arr.removeAt(0);
  }

  List<SeriesPoint> seriesFor(String key) => series[key] ?? const [];

  void _upsertAction(Map<String, dynamic> a) {
    final id = _asInt(a['id'])!;
    final m = actions.putIfAbsent(id, () {
      actionOrder.insert(0, id);
      return ActionCard(id);
    });
    m.cause = a['cause'] as String? ?? m.cause;
    m.command = a['command_topic'] as String? ?? m.command;
    if (a['evidence'] != null) m.evidence = (a['evidence'] as Map).cast<String, dynamic>();
    final expectation = a['expectation'];
    if (expectation != null) m.expect = (expectation as Map)['describe'] as String?;
    if (a['retries'] != null) m.retries = _asInt(a['retries'])!;
  }

  void _markAction(int id, String status, {String? reason}) {
    final m = actions.putIfAbsent(id, () {
      actionOrder.insert(0, id);
      return ActionCard(id);
    });
    m.status = status;
    if (reason != null) m.reason = reason;
  }

  dynamic point(String key, [dynamic dflt]) => points[key]?.value ?? dflt;

  void reset() {
    now = 0;
    points.clear();
    series.clear();
    actions.clear();
    actionOrder.clear();
    feed.clear();
    notifyListeners();
  }

  static int? _asInt(dynamic v) => v == null ? null : (v as num).toInt();
}
